pymeasure/instruments/activetechnologies/AWG401x.py

```diff
@@ -0,0 +1,1106 @@
+        """This property sets or queries the sample rate for the Sampling
+        Clock.""",
+        validator=strict_range,
+        dynamic=True
+    )
+
+    sampling_rate_max = Instrument.measurement(
+        "AWGControl:SRATe? MAXimum",
+        """This property queries the maximum sample rate for the Sampling
+        Clock."""
+    )
+
+    sampling_rate_min = Instrument.measurement(
+        "AWGControl:SRATe? MINimum",
+        """This property queries the minimum sample rate for the Sampling
+        Clock."""
+    )
+
+    run_status = Instrument.measurement(
+        "AWGControl:RSTATe?",
+        """This property returns the run state of the AWG. The possible values
+        are: STOPPED, WAITING_TRIGGER, RUNNING""",
+        values={"STOPPED": 0, "WAITING_TRIGGER": 1, "RUNNING": 2},
+        map_values=True
+    )
+
+    enabled = Instrument.control(
+        "AWGControl:RSTATe?", "AWGControl:%s",
+        """A boolean property that enables the generation of signals.""",
+        validator=strict_discrete_set,
+        values={True: "RUN", False: "STOP"},
+        map_values=True,
+        get_process=lambda v: "STOP" if v == 0 else "RUN"
+    )
+
+    trigger_source = Instrument.control(
+        "TRIGger:SEQuence:SOURce?", "TRIGger:SEQuence:SOURce %s",
+        """This property sets or returns the instrument trigger source. The
+        possible values are:
+
+        * TIM<ER>: the trigger is sent at regular intervals.
+        * EXT<ERNAL>: the trigger come from the external BNC connector.
+        * MAN<UAL>: the trigger is sent via software or using the trigger
+          button on front panel.""",
+        validator=strict_discrete_set,
+        values=["TIMER", "TIM", "EXTERNAL", "EXT", "MANUAL", "MAN"]
+    )
+
+    waveforms = property(
+        lambda self: self._waveforms,
+        doc="""This property returns a dict with all the waveform present
+        in the instrument system (Wave. List). It is possible to modify the
+        values, delete them or create new waveforms""")
+
+    def __init__(self, adapter, **kwargs):
+        super().__init__(adapter, **kwargs)
+
+        self.setting_ch = {}
+        for i in range(1, self.num_ch+1):
+            self.setting_ch[i] = ChannelAWG(self, i)
+
+        self.entries = self.DummyEntriesElements(self, self.num_ch)
+        # self.sequencer = Sequencer(self, self.num_ch)
+        self.burst_count_values = list(map(
+            int,
+            [self.burst_count_min, self.burst_count_max]))
+
+        self.sampling_rate_values = list(map(
+            float,
+            [self.sampling_rate_min, self.sampling_rate_max]))
+
+        self._waveforms = self.WaveformsLazyDict(self)
+
+    def trigger(self):
+        """Force a trigger event to occour."""
+        self.write("TRIGger:SEQuence:IMMediate")
+
+    def save_file(self,
+                  file_name: str,
+                  data,
+                  path: str = None,
+                  override_existing=False):
```

Comment:
type hinting is not encouraged in `PyMeasure`, at least from some discussion dated back in 2020 (see #259).

Reply:
type hinting removed